### ISSUE TITLE
HEVC decoding on Raspberry Pi 4

### ISSUE THREAD

**Issue by michellabbe** [moonlight-stream/moonlight-qt#409]
More apps are getting support for HEVC hardware decoding on the RPi4, including [Steam Link](https://steamcommunity.com/app/353380/discussions/6/3681064390585946947/), and of course that would be a great addition (+ 4K support) to Moonlight.

Current status of software support for hardware HEVC decoding on RPi4 is still very confusing.

From what I understand, Kodi, LibreELEC and info-beamer used some custom patched ffmpeg that produces data usable in MMAL buffers (even though MMAL doesn't and will never support h.265).  Later, the Raspbian FFMPEG added some other patch for HEVC support, which is what VLC is using.

I'm not sure if the final target revision of the HEVC decoder is completed by now (6 months ago it was supposedly waiting on some kernel support using V4L2, see links below), but V4L2 seems to be available in the current FFMPEG version:

```
pi@rpi4:~ $ ffmpeg -decoders | grep v4l2
ffmpeg version 4.1.4-1+rpt7~deb10u1 Copyright (c) 2000-2019 the FFmpeg developers
  built with gcc 8 (Raspbian 8.3.0-6+rpi1)
[...]
 V..... h263_v4l2m2m         V4L2 mem2mem H.263 decoder wrapper (codec h263)
 V..... h264_v4l2m2m         V4L2 mem2mem H.264 decoder wrapper (codec h264)
 V..... hevc_v4l2m2m         V4L2 mem2mem HEVC decoder wrapper (codec hevc)
 V..... mpeg1_v4l2m2m        V4L2 mem2mem MPEG1 decoder wrapper (codec mpeg1video)
 V..... mpeg2_v4l2m2m        V4L2 mem2mem MPEG2 decoder wrapper (codec mpeg2video)
 V..... mpeg4_v4l2m2m        V4L2 mem2mem MPEG4 decoder wrapper (codec mpeg4)
 V..... vc1_v4l2m2m          V4L2 mem2mem VC1 decoder wrapper (codec vc1)
 V..... vp8_v4l2m2m          V4L2 mem2mem VP8 decoder wrapper (codec vp8)
 V..... vp9_v4l2m2m          V4L2 mem2mem VP9 decoder wrapper (codec vp9)
```

See these threads:
https://www.raspberrypi.org/forums/viewtopic.php?t=252053
https://www.reddit.com/r/raspberry_pi/comments/e64bl6/is_the_4bs_h265_hardware_decode_currently/

and also:
[https://github.com/raspberrypi/firmware/issues/1168](https://github.com/raspberrypi/firmware/issues/1168)

**Comment by cgutman** [comment]
The existing V4L2 decoders in FFmpeg don't work for the Pi's HEVC V4L2 decoder is stateless. We would have to use one of the custom patched FFmpeg builds.

https://www.raspberrypi.org/forums/viewtopic.php?t=267555#p1624571

**Comment by Vliro** [comment]
I saw this issue and wondering if it is possible to accomplish this via e.g. compiling ffmpeg using the patches stated above or if it is an integral part of moonlight-qt and not easily changed?

**Comment by cpetry** [comment]
I'm also quite interested on this.
Is rpi 4 with raspbian buster not (yet) capable of this? Think I read something about this recently...

**Comment by david97j** [comment]
Chromium has just recieved hardware decoding for the pi 4, i think it should be posible to implement in moonlight and it would be awesome

**Comment by SupervisedThinking** [comment]
@cgutman 
Maybe you could make use of some LibreELEC FFmpeg patches again?

https://github.com/LibreELEC/LibreELEC.tv/tree/master/packages/multimedia/ffmpeg/patches
https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/multimedia/ffmpeg/package.mk#L22-L57

**Comment by cgutman** [comment]
Here is a build with HEVC support: https://ci.appveyor.com/api/buildjobs/k7i066nuvgodqied/artifacts/moonlight-qt_3.1.1-1_armhf.deb

You must add `dtoverlay=rpivid-v4l2` line to your `/boot/config.txt` and reboot for the HEVC codec to be enabled.

Let me know how it works for you.

**Comment by michellabbe** [comment]
Yay! :)

I added the line to `/boot/config.txt`, rebooted and installed it on top of standard v3.1.1.

At first it wouldn't open but `moonlight-qt` is hinting properly at enabling the 'Fake KMS' driver in `raspi-config`.

Is Performance Stats Overlay (Ctrl+Alt+Shift+S) supposed to work on this version?
(already reported [here](https://github.com/moonlight-stream/moonlight-qt/issues/160#issuecomment-585767116) but you didn't confirm, also I've seen #405 but it mentions Archlinux and not specifically RPi OS).

I'm also seeing a couple `StackView has detected conflicting anchors. Transitions may not execute properly` warnings in the logs. Are these normal?

See [moonlight-qt.log](https://github.com/moonlight-stream/moonlight-qt/files/6412438/moonlight-qt.log)

Streaming with hardware HEVC decoding is working fine so far though, and the UI doesn't seem to suffer tearing when launched in console.

**Comment by cgutman** [comment]
> Is Performance Stats Overlay (Ctrl+Alt+Shift+S) supposed to work on this version?
> (already reported [here](https://github.com/moonlight-stream/moonlight-qt/issues/160#issuecomment-585767116) but you didn't confirm, also I've seen #405 but it mentions Archlinux and not specifically RPi OS).

Nope, it's not expected to work. Until the Raspberry Pi supports importing DRM buffers from the HEVC codec (and does so with good performance), this will not be supported (more details [here](https://github.com/moonlight-stream/moonlight-qt/issues/553#issuecomment-804638387)). Personally I'm not holding my breath. Other SoCs like RK3389 can handle this just fine and it "[just works](https://github.com/moonlight-stream/moonlight-qt/issues/509#issuecomment-787488648)" with good performance. To work around the awful GL performance on the Pi, we have to render using the DRM API directly, which means no overlays.
 
> I'm also seeing a couple `StackView has detected conflicting anchors. Transitions may not execute properly` warnings in the logs. Are these normal?

Yes, this is normal for the older Qt version that Debian 10 has.

> See [moonlight-qt.log](https://github.com/moonlight-stream/moonlight-qt/files/6412438/moonlight-qt.log)
> 
> Streaming with hardware HEVC decoding is working fine so far though, and the UI doesn't seem to suffer tearing when launched in console.

Nice!

**Comment by cgutman** [comment]
Closing as this is implemented in v3.1.2
